radicle-tracker/src/thread.rs

```diff
@@ -0,0 +1,827 @@
+    #[error("Tried to move to previous item in the main thread, but we are at the first")]
+    PreviousMainOutOfBounds,
+    /// An attempt was made to move to the previous item in a reply thread, but
+    /// the pointer is on the main thread.
+    #[error("Cannot move to previous item in a thread when we are located on the main thread")]
+    PreviousThreadOnMain,
+    /// An attempt was made to move to the previous item in the main thread, but
+    /// the pointer is already at the last item.
+    #[error("Tried to move to next item in the main thread, but we are at the last")]
+    NextMainOutOfBounds,
+    /// An attempt was made to move to the next item in a reply thread, but
+    /// the pointer is already at the last item.
+    #[error("Tried to move to next item in the reply thread, but we are at the last")]
+    NextRepliesOutOfBound,
+    /// An attempt was made to delete the root of a thread, but the root is
+    /// immutable.
+    #[error("Cannot delete the main item of the thread")]
+    DeleteRoot,
+}
+
+/// A collection of replies where a reply is any item that has a [`Status`].
+///
+/// `Replies` are deliberately opaque as they should mostly be interacted with
+/// via [`Thread`].
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Replies<A>(NonEmpty<Status<A>>);
+
+impl<A> Replies<A> {
+    fn new(a: A) -> Self {
+        Replies(NonEmpty::new(Status::Live(a)))
+    }
+
+    fn reply(&mut self, a: A) {
+        self.0.push(Status::Live(a))
+    }
+
+    fn first(&self) -> &Status<A> {
+        self.0.first()
+    }
+
+    fn first_mut(&mut self) -> &mut Status<A> {
+        self.0.first_mut()
+    }
+
+    /// Check is the `Replies` are empty. It is always `false`.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Check the length of the `Replies`.
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    fn get(&self, index: usize) -> Option<&Status<A>> {
+        self.0.get(index)
+    }
+
+    fn get_mut(&mut self, index: usize) -> Option<&mut Status<A>> {
+        self.0.get_mut(index)
+    }
+
+    /// Get the [`Iterator`] for the `Replies`.
+    pub fn iter<'a>(&'a self) -> impl Iterator<Item = &Status<A>> + 'a {
+        self.0.iter()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum Finger {
+    Root,
+    Main(usize),
+    Thread((usize, usize)),
+}
+
+// This point to the main thread, and the first item in that thread.
+const ROOT_FINGER: Finger = Finger::Root;
+
+/// A `Thread` is the root item followed by a series of non-empty replies to the
+/// root item. For each item in reply to the root item there may be 0 or more
+/// replies.
+#[derive(Debug, Clone)]
+pub struct Thread<A> {
+    // A finger points into the `main_thread` structure.
+    // If it is `Left` then it is pointing to the main thread.
+    // If it is `Right` then it is pointing to a reply to a comment in the main thread.
+    _finger: Finger,
```

Comment:
What's the reason this is prefixed with `_`? Normally that means it's unused..

Reply:
I wanted to convey that this is only something the data structure carries around and uses internally but has nothing to do with the actual data itself. Does that make sense? Should I remove the `_`?

Reply:
Ah I see. I would remove it still. Maybe distinguish it from the other fields with a blank line? `_finger` is ignored by the compiler if unused, it has special semantics.